src/Compilers/VisualBasic/Portable/CodeGen/ResumableStateMachineStateAllocator.vb

```diff
@@ -0,0 +1,91 @@
+    ''' Allocates resumable states, i.e. states that resume execution of the state machine after await expression or yield return.
+    ''' </summary>
+    Friend NotInheritable Class ResumableStateMachineStateAllocator
+        Private ReadOnly _slotAllocator As VariableSlotAllocator
+        Private ReadOnly _increasing As Boolean
+        Private ReadOnly _firstState As Integer
+
+        ''' <summary>
+        ''' The number of the next generated resumable state (i.e. state that resumes execution of the state machine after await expression or yield return).
+        ''' </summary>
+        Private _nextState As Integer
+
+#If DEBUG Then
+        ''' <summary>
+        ''' EnC support: states in this state machine that were matched to states of the previous generation state machine.
+        ''' </summary>
+        Private _matchedStates As BitVector = BitVector.Empty
+#End If
+        ''' <summary>
+        ''' EnC support: number of states in this state machine that match states of the previous generation state machine.
+        ''' </summary>
+        Private _matchedStateCount As Integer
+
+        Public Sub New(slotAllocator As VariableSlotAllocator, firstState As Integer, increasing As Boolean)
+            _increasing = increasing
+            _slotAllocator = slotAllocator
+            _matchedStateCount = 0
+            _firstState = firstState
+            _nextState = If(slotAllocator?.GetFirstUnusedStateMachineState(increasing), firstState)
+        End Sub
+
+        Public Function AllocateState(awaitOrYieldReturnSyntax As SyntaxNode) As Integer
+            Debug.Assert(SyntaxBindingUtilities.BindsToResumableStateMachineState(awaitOrYieldReturnSyntax))
+
+            Dim direction = If(_increasing, +1, -1)
+            Dim stateNumber As Integer
+
+            If _slotAllocator?.TryGetPreviousStateMachineState(awaitOrYieldReturnSyntax, stateNumber) = True Then
+#If DEBUG Then
+                ' two states of the New state machine should Not match the same state of the previous machine
+                Debug.Assert(Not _matchedStates(stateNumber * direction))
+                _matchedStates(stateNumber * direction) = True
+#End If
+                _matchedStateCount += 1
+            Else
+                stateNumber = _nextState
+                _nextState += direction
+            End If
+
+            Return stateNumber
+        End Function
+
+        ''' <summary>
+        ''' True if any of the states generated for any previous state machine has not been allocated in this version.
+        ''' </summary>
+        Public ReadOnly Property HasMissingStates As Boolean
+            Get
+                Return _matchedStateCount < If(_slotAllocator?.GetFirstUnusedStateMachineState(_increasing), _firstState) - _firstState
+            End Get
+        End Property
+
+        Public Function GenerateThrowMissingStateDispatch(f As SyntheticBoundNodeFactory, cachedState As BoundExpression, message As String) As BoundStatement
```

Comment:
Given that the statement this returns will be injected at then end of the dispatch, why do we need to condition it in any way? Why not just return a `throw new InvalidOperationException(message)` without the enclosing `if`?

Reply:
We need to condition. The dispatch falls through for initial state. Without the condition we would throw for initial state as well.